### PR TITLE
blockchain: enforce 10 block age for spending outputs

### PR DESCRIPTION
**Pull request by moneromooo-monero**
Some custom wallet code apparently ignores this, which causes users
of that code to be fingerprinted

### CODE REVIEW

File: tests/core_tests/rct.cpp [hunk 1]
```diff
@@ -507,3 +517,11 @@ bool gen_rct_tx_rct_altered_extra::generate(std::vector<test_event_entry>& event
     NULL, [&failed](transaction &tx) {std::string extra_nonce; crypto::hash pid = crypto::null_hash; set_payment_id_to_tx_extra_nonce(extra_nonce, pid); if (!add_extra_nonce_to_tx_extra(tx.extra, extra_nonce)) failed = true; }) && !failed;
 }
 
+bool gen_rct_tx_uses_output_too_early::generate(std::vector<test_event_entry>& events) const
+{
+  const int mixin = 10;
+  const int out_idx[] = {1, -1};
+  const uint64_t amount_paid = 10000;
+  const rct::RCTConfig rct_config { rct::RangeProofPaddedBulletproof, 2 };
+  return generate_with_full(events, out_idx, mixin, amount_paid, 7, 12, rct_config, false, NULL, NULL);
```
Comment (stoffu) [inline tests/core_tests/rct.cpp]: Maybe use `CRYPTONOTE_DEFAULT_TX_SPENDABLE_AGE - 3` and `HF_VERSION_ENFORCE_MIN_AGE` to clarify the intent?

File: tests/core_tests/rct.cpp [hunk 2]
```diff
@@ -214,6 +214,7 @@ bool gen_rct_tx_validation_base::generate_with(std::vector<test_event_entry>& ev
   td.amount = amount_paid;
   std::vector<tx_destination_entry> destinations;
   destinations.push_back(td);
+  destinations.push_back(td);
```
Comment (stoffu) [inline tests/core_tests/rct.cpp]: Maybe add a comment mentioning the new rule about the minimum 2 outs?

File: tests/core_tests/rct.cpp [hunk 3]
```diff
@@ -237,6 +239,15 @@ bool gen_rct_tx_validation_base::generate_with(std::vector<test_event_entry>& ev
   return true;
 }
 
+bool gen_rct_tx_validation_base::generate_with(std::vector<test_event_entry>& events,
+    const int *out_idx, int mixin, uint64_t amount_paid, bool valid,
+    const std::function<void(std::vector<tx_source_entry> &sources, std::vector<tx_destination_entry> &destinations)> &pre_tx,
+    const std::function<void(transaction &tx)> &post_tx) const
+{
+  const rct::RCTConfig rct_config { rct::RangeProofBorromean, 0 };
+  return generate_with_full(events, out_idx, mixin, amount_paid, 10, 4, rct_config, valid, pre_tx, post_tx);
```
Comment (stoffu) [inline tests/core_tests/rct.cpp]: Another bikeshedding: `10` could be written as `CRYPTONOTE_DEFAULT_TX_SPENDABLE_AGE`

**Comment by 1blockologist** [comment]
can we make it lower than 10 blocks if we must enforce this on a consensus level?

how about 5 or 7? the sets are larger now and have capacity to grow much larger given the lower transaction size improvements

**Comment by fullmetalScience** [comment]
I feel this topic cannot be taken lightly and should receive more discussion - especially on a scientific level - in order to find an optimal solution or at least a compromise that allows for certain real-world adoption for Monero as digital, every-day cash.

As a base for discussion, please find [my recent use-case example](https://www.reddit.com/r/Monero/comments/d0x633/before_10_confirmations_is_enforced_by_the/eztqm84/) over at the Monero Sub-Reddit.

Are there any related MRL resources that discuss the issue?

**Comment by who-biz** [comment]
Does this actually prevent spending of outputs prior to `height + 10`?  It seems an awful lot like one could simply change the `HF_VERSION_ENFORCE_MIN_AGE` to `14` (or some arbitarily higher number than `13`) and `CRYPTONOTE_DEFAULT_TX_SPENDABLE_AGE` to `0` (or some arbitrarily lower number than `10`) in `src/cryptonote_config.h`... and the transaction would still go through, no? 

Same method, as I described here: From the discussion here: https://github.com/monero-project/monero/issues/5712, pretty much. 

There was a check like this in `blockchain.cpp`, around July 24th (this has since vanished) and that did nothing to actually impact consensus... So, I'd like to hear some reasoning in support of why the addition of a different pre-compile constant would be any different. It does the same thing, functionally, it appears. (i.e. `nothing` in terms of consensus)

**Comment by SamsungGalaxyPlayer** [comment]
@1blockologist I think a discussion regarding the number of blocks is out of scope for here. Open a new issue or find a relevant one.

**Comment by 1blockologist** [comment]
@SamsungGalaxyPlayer lets enforce a 5 block age for spending outputs

and let wallets enforce a higher one if they want.

**Comment by dEBRUYNE-1** [comment]
@stoffu - Can you perhaps give this another approval after the rebase?

**Pull request merged**